### ISSUE TITLE
How can I restart or resume interrupted hyperparameter evolution?

### ISSUE THREAD

**Issue by erb13020** [ultralytics/yolov5#5996]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I would like to perform hyperparameter evolution to optimize my model. According to my calculations, this will take about 1 month of training time for 300 generations.

I live in an area where there maybe 1 or 2 power outages in a month. If I am performing evolution and it is interrupted by a power outage, is there any way to resume the evolution where I left off? Should I continue the process using either the latest or best hyperparameters?

### Additional

I was thinking of purchasing a backup power supply for my desktop, as these outages are usually less than 10 minutes.

**Comment by glenn-jocher** [comment]
@erb13020 👋 Hello! Thanks for asking about **resuming evolution**. Yes a UPS would be a good investment in your situation.

Resuming YOLOv5 🚀 evolution is a bit different than resuming a normal training run with `python train.py --resume`. If you started an evolution run which was interrupted, or finished normally, and you would like to continue for additional generations where you left off, then you pass `--resume` and specify the `--name` of the evolution you want to resume, i.e.:


## Start Evolution

Assume you evolve YOLOv5s on COCO128 for 2 epochs for 3 generations:
```shell
python train.py --epochs 10 --data coco128.yaml --weights yolov5s.pt --evolve 3
```
If this is your first evolution a new directory `runs/evolve/exp` will be created to save your results.
```shell
# ├── yolov5
#     └── runs
#         └── evolve
#             └── exp  ← evolution saved here
```

## Start a Second Evolution

Now assume you want to start a completely separate evolution: YOLOv5s on VOC for 5 epochs for 3 generations. You simply start evolving, and your new evolution will again be logged to a new directory `runs/evolve/exp2`:
```shell
python train.py --epochs 5 --data VOC.yaml --weights yolov5s.pt --evolve 3
```

You will now have two evolution runs saved:
```shell
# ├── yolov5
#     └── runs
#         └── evolve
#             ├── exp  ← first evolution (COCO128)
#             └── exp2  ← second evolution (VOC)
```

Notebook example: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
<img width="1396" alt="Screenshot 2021-09-15 at 12 23 13" src="https://user-images.githubusercontent.com/26833433/133416922-59a3e9f5-2630-4875-9c64-a982e312b400.png">


## Resume an Evolution

If you want to resume the first evolution (COCO128 saved to `runs/evolve/exp`), regardless of whether it was interrupted or finished successfully, then you use the same exact command you started with plus `--resume --name exp`, passing the **additional** number of generations you want, i.e. `--evolve 30` for 30 more generations:
```shell
python train.py --epochs 10 --data coco128.yaml --weights yolov5s.pt --evolve 30 --resume --name exp
```
Evolution will run for an additional 30 generations and all new results will be appended to the existing `runs/evolve/exp/evolve.csv`. 

Good luck 🍀 and let us know if you have any other questions!

**Comment by unrue** [comment]
> ## Resume an Evolution
> If you want to resume the first evolution (COCO128 saved to `runs/evolve/exp`), regardless of whether it was interrupted or finished successfully, then you use the same exact command you started with plus `--resume --name exp`, passing the **additional** number of generations you want, i.e. `--evolve 30` for 30 more generations:
> 
> ```shell
> python train.py --epochs 10 --data coco128.yaml --weights yolov5s.pt --evolve 30 --resume --name exp
> ```

> Good luck 🍀 and let us know if you have any other questions!

Hi, "--cache" parameter is not necessary for resume? Thanks.

**Comment by glenn-jocher** [comment]
@unrue If you previously used `--cache` during evolution, you should include it again when resuming. Good luck with your evolutions, and feel free to ask if you have more questions! 🚀

**Comment by unrue** [comment]
Hi Glenn, may I resume also multigpu evolve phase?

**Comment by glenn-jocher** [comment]
Hello @unrue,

Yes, you can resume a multi-GPU evolution phase as well. Just ensure you use the same command structure you initially used, including `--resume` and the appropriate `--name` for the evolution run. Here's an example:

```shell
for i in 0 1 2 3 4 5 6 7; do
  sleep $(expr 30 \* $i) &&  # 30-second delay (optional)
  echo 'Starting GPU '$i'...' &&
  nohup python train.py --epochs 10 --data coco128.yaml --weights yolov5s.pt --cache --device $i --evolve 30 --resume --name exp > evolve_gpu_$i.log &
done
```

This will resume the evolution on multiple GPUs. Good luck with your experiments! 🚀

**Comment by unrue** [comment]
Thanks Glenn, last little question:

The evolve phase, should be done before the final training or after?

I mean, through the evolve phase I'm able to find best hyper-parameters. Such parameters, should be used to make a final training, or not?

In evolve directory, I don't see anything like weights/best.pt:

results.csv results.png hyp_evolve.yaml evolve.csv labels_correlogram.jpg train_batch1.jpg train_batch0.jpg labels.jpg

So, I suppose that once the evolve phase is finished, I'm able to train my net using best hyperparameter possible with hyp_evolve.yaml. Right? Thanks.

**Comment by glenn-jocher** [comment]
Hello!

You're correct in your understanding. The evolve phase is designed to help you find the best hyperparameters before you conduct your final training. Once the evolution is complete, you should use the `hyp_evolved.yaml` file, which contains the optimized hyperparameters, for your final training session. This approach ensures that your model is trained with the most effective settings discovered during the evolution.

Indeed, the evolve directory does not store model weights like `
